### PR TITLE
[Bugfix][Structured Output] Let terminal grammars stop under min_tokens

### PR DESCRIPTION
**Pull request by Sunt-ing**
## Purpose

`min_tokens` can force structured-output requests to sample outside the grammar after a short schema reaches its terminal state. For schemas such as `{"type":"boolean"}`, `{"enum":["a"]}`, or `{"type":"null"}`, the grammar's only legal continuation after the value is EOS. `min_tokens` then masks EOS because the request has not reached the requested minimum length, leaving no legal token.

On xgrammar this returns `finish_reason="error"` with a trailing `!`, for example `true!`. On guidance the same conflict silently emits `!` until `min_tokens` is satisfied. This PR lets stop/EOS survive only when the `min_tokens` mask would otherwise empty the row, then finishes the request once the structured grammar accepts its terminal token. Normal `min_tokens` behavior is unchanged when any non-stop token remains available, and the same guard is forwarded through the V2 spec-decode sampler path.

This is not covered by the existing FSM-adjacent work I found: #45015 is pipeline-parallel stale FSM state, #44297 and #44993 are spec-decode reasoning-boundary masking, #37506 is `ignore_eos` plus extra speculative tokens, and #42619 is an already-terminated matcher being advanced later. I re-ran current `main` after #44297 and it still fails the `min_tokens` terminal-schema case below; none of the adjacent fixes handles `min_tokens` masking the only grammar-legal EOS token.

## Test Plan

Run a real `LLM.generate` A/B on Qwen3-0.6B with xgrammar structured output. The before case uses the PR diff against upstream `main` reversed.

Also run the focused V1 regression test, the logits-processor correctness suite, a CUDA smoke for the GPU all-masked fallback, ruff on changed files, and `git diff --check`.

## Test Result

Qwen3-0.6B A/B flips the bug on current `main` after #44297: with the patch, boolean/enum/null terminal schemas stop cleanly at EOS; with the patch reversed, the same requests emit `!` and finish with `error`. Guidance shows the same boolean conflict: before it emits `true!!!!!!!!!!!!!!!!!!!!`, after it returns `true` and stops at EOS.

Focused tests and checks passed:

```text
tests/v1/e2e/general/test_min_tokens.py::test_min_tokens_with_terminal_structured_output
1 passed, 22 warnings in 5.68s

tests/v1/logits_processors/test_correctness.py::test_logitsprocs
6 passed, 22 warnings in 109.19s

GPU all-masked fallback smoke
PASS

ruff check changed files
All checks passed!

git diff --check
PASS
```

<details>
<summary>Qwen3-0.6B A/B reproduction</summary>

Environment:

```text
GPU: NVIDIA GeForce RTX 4090
model: Qwen/Qwen3-0.6B, snapshot c1899de289a04d12100db370d81485cdf75e47ca
backend: xgrammar
env: HF_HUB_OFFLINE=1 TRANSFORMERS_OFFLINE=1 VLLM_ENABLE_V1_MULTIPROCESSING=0 VLLM_USE_FLASHINFER_SAMPLER=0
```

Command:

```bash
cd /path/to/vllm
export PYTHONPATH="$PWD"
export VLLM_ENABLE_V1_MULTIPROCESSING=0
export VLLM_USE_FLASHINFER_SAMPLER=0
export HF_HUB_OFFLINE=1
export TRANSFORMERS_OFFLINE=1
export CUDA_VISIBLE_DEVICES=0

cat >/tmp/min_tokens_terminal_so.py <<'PY'
import json
import os

from vllm import LLM, SamplingParams
from vllm.sampling_params import StructuredOutputsParams

MODEL = os.environ.get("VLLM7_MODEL", "/root/.cache/huggingface/hub/models--Qwen--Qwen3-0.6B/snapshots/c1899de289a04d12100db370d81485cdf75e47ca")

def run(llm, label, schema, min_tokens):
    params = SamplingParams(
        temperature=0.0,
        min_tokens=min_tokens,
        max_tokens=40,
        structured_outputs=StructuredOutputsParams(json=json.dumps(schema)),
    )
    output = llm.generate(["Answer:"], params, use_tqdm=False)[0].outputs[0]
    print(label, repr(output.text), len(output.token_ids), output.token_ids, output.finish_reason, output.stop_reason)

llm = LLM(
    model=MODEL,
    enforce_eager=True,
    gpu_memory_utilization=0.30,
    max_model_len=512,
    structured_outputs_config={"backend": "xgrammar"},
)
run(llm, "control-bool", {"type": "boolean"}, 0)
run(llm, "min-bool", {"type": "boolean"}, 20)
run(llm, "min-enum", {"enum": ["a"]}, 20)
run(llm, "min-null", {"type": "null"}, 30)
PY

python /tmp/min_tokens_terminal_so.py

git fetch https://github.com/vllm-project/vllm.git main
BASE=$(git merge-base FETCH_HEAD HEAD)
git diff "$BASE"..HEAD >/tmp/min_tokens_terminal_so.patch
git apply -R /tmp/min_tokens_terminal_so.patch
python /tmp/min_tokens_terminal_so.py
git apply /tmp/min_tokens_terminal_so.patch
```

After, with this PR:

```text
control-bool 'true' 3 [376, 361, 151645] stop None
min-bool 'true' 3 [376, 361, 151645] stop None
min-enum '"a"' 4 [1, 64, 1, 151645] stop None
min-null 'null' 4 [77, 360, 75, 151645] stop None
```

Before, with this patch reversed:

```text
control-bool 'true' 3 [376, 361, 151645] stop None
ERROR backend_xgrammar.py: Failed to advance FSM for request ... for tokens 0. Please file an issue.
ERROR scheduler.py: Unexpected: grammar rejected tokens [0] for request ... Terminating request.
min-bool 'true!' 3 [376, 361, 0] error None
ERROR backend_xgrammar.py: Failed to advance FSM for request ... for tokens 0. Please file an issue.
ERROR scheduler.py: Unexpected: grammar rejected tokens [0] for request ... Terminating request.
min-enum '"a"!' 4 [1, 64, 1, 0] error None
ERROR backend_xgrammar.py: Failed to advance FSM for request ... for tokens 0. Please file an issue.
ERROR scheduler.py: Unexpected: grammar rejected tokens [0] for request ... Terminating request.
min-null 'null!' 4 [77, 360, 75, 0] error None
```

</details>

AI assistance was used to prepare this PR.

cc @mgoin @njhill

### CODE REVIEW

File: vllm/v1/sample/logits_processor/builtin.py [hunk 1]
```diff
@@ -226,10 +226,26 @@ def update_state(self, batch_update: BatchUpdate | None):
...
 
+    def _mask_stop_token_logits(
+        self,
+        logits: torch.Tensor,
+        logits_slice: tuple[torch.Tensor, torch.Tensor],
+    ) -> None:
+        stop_logits = logits[logits_slice].clone()
+        logits.index_put_(logits_slice, self.neg_inf_tensor)
+
+        empty_rows = torch.isneginf(logits).all(dim=-1)
```
Comment (arpera) [inline vllm/v1/sample/logits_processor/builtin.py]: Can this additional check impact performance of requests with min_tokens set but without structured output?

Comment (Sunt-ing) [reply vllm/v1/sample/logits_processor/builtin.py]: Thanks @arpera, good catch. Yes, the generic path could run the all-masked check for non-structured-output `min_tokens` requests. I scoped the fallback to requests with `structured_outputs`; on the V2 GPU path, batches without structured output keep the constexpr fast path, and mixed batches use a per-request gate so ordinary `min_tokens` rows do not run the all-vocab restore scan. Verified with a mixed GPU smoke plus the logits processor tests.

File: vllm/v1/sample/logits_processor/builtin.py [hunk 2]
```diff
@@ -168,14 +168,19 @@ def __init__(
     ):
         # index -> (min_toks, output_token_ids, stop_token_ids)
```
Comment (arpera) [inline vllm/v1/sample/logits_processor/builtin.py]: This comment needs to be updated then

File: vllm/v1/sample/logits_processor/builtin.py [hunk 3]
```diff
@@ -214,22 +224,58 @@ def update_state(self, batch_update: BatchUpdate | None):
...
         if self.min_toks:
-            # Inhibit EOS token for requests which have not reached min length
-            logits.index_put_(self.logits_slice, self.neg_inf_tensor)
+            # Inhibit EOS token for requests which have not reached min length.
+            self._mask_stop_token_logits(
+                logits, self.logits_slice, self.preserve_logits_slice
+            )
         return logits
 
     def apply_with_spec_decode(
```
Comment (arpera) [inline vllm/v1/sample/logits_processor/builtin.py:288]: Maybe I missed something in the patch but do we cover `apply_with_spec_decode` with any unit tests?

File: vllm/v1/sample/logits_processor/builtin.py [hunk 4]
```diff
@@ -214,22 +224,58 @@ def update_state(self, batch_update: BatchUpdate | None):
...
+        preserve_logits_slice: tuple[torch.Tensor, torch.Tensor],
+    ) -> None:
+        if preserve_logits_slice[0].numel() == 0:
+            logits.index_put_(logits_slice, self.neg_inf_tensor)
+            return
+
+        stop_logits = logits[preserve_logits_slice].clone()
+        logits.index_put_(logits_slice, self.neg_inf_tensor)
+
+        restore_mask = torch.isneginf(logits[preserve_logits_slice[0]]).all(dim=-1)
```
Comment (arpera) [inline vllm/v1/sample/logits_processor/builtin.py]: In this case `preserve_logits_slice` can contain lines duplicates in case request has more than 1 stop token. Can this impact performance?

File: vllm/v1/sample/logits_processor/builtin.py [hunk 5]
```diff
@@ -214,22 +224,58 @@ def update_state(self, batch_update: BatchUpdate | None):
...
+        restore_slice = (
+            preserve_logits_slice[0][restore_mask],
+            preserve_logits_slice[1][restore_mask],
+        )
+        logits.index_put_(restore_slice, stop_logits[restore_mask])
```
Comment (arpera) [inline vllm/v1/sample/logits_processor/builtin.py:275]: Can we guarantee here that logits is not all of -inf now? I cannot say it for sure.

File: vllm/v1/sample/logits_processor/builtin.py (hunk 2 above)
Comment (Sunt-ing) [reply vllm/v1/sample/logits_processor/builtin.py]: Updated

File: vllm/v1/sample/logits_processor/builtin.py (hunk 3 above)
Comment (Sunt-ing) [reply vllm/v1/sample/logits_processor/builtin.py:288]: Added one focused `apply_with_spec_decode` test for the structured-output case

File: vllm/v1/sample/logits_processor/builtin.py (hunk 4 above)
Comment (Sunt-ing) [reply vllm/v1/sample/logits_processor/builtin.py]: Changed this to check each unique row once, then map the result back to the preserved stop-token entries

File: vllm/v1/sample/logits_processor/builtin.py (hunk 5 above)
Comment (Sunt-ing) [reply vllm/v1/sample/logits_processor/builtin.py:275]: Right, if a preserved stop-token logit was already `-inf`, there is nothing safe to restore. The fallback now only restores finite stop logits captured before `min_tokens` masking.

File: vllm/v1/core/sched/scheduler.py [hunk 6]
```diff
@@ -1646,18 +1646,20 @@ def update_from_output(
...
+                        logger.error(
+                            "Unexpected: grammar rejected tokens %s for request %s. "
+                            "Terminating request.",
+                            advance_token_ids,
+                            req_id,
+                        )
+                        request.status = RequestStatus.FINISHED_ERROR
+                        request.resumable = False
+                        stopped = True
+                    elif not stopped and grammar.is_terminated():
```
Comment (arpera) [inline vllm/v1/core/sched/scheduler.py]: I have a question about what actually means that grammar is in terminated state. For example, let us consider simple schema `{"type":"boolean"}`. Model is going to generate `true<EOS>` sequence, so we have two tokens `true` and `<EOS>`.

Is it possible that after accepting token `true` grammar will be in terminated state already without accepting next `<EOS>` token? I am not sure about it since we can use three different grammar backends here (xgrammar, lguidance, outlines) and I didn't check that every backend does not terminate before `<EOS>` token.

Instead of checking every backend's source code (which may obviously change in the future) I would propose to put there some kind of assertion checking that the last accepted token was `<EOS>`. I don't know how to implement it neatly there but this is just my thoughts on what might go wrong if we introduce this new logic.

File: vllm/v1/worker/gpu/sample/logit_bias.py [hunk 7]
```diff
@@ -232,11 +245,37 @@ def _bias_kernel(
...
         tl.store(
             logits_ptr + token_idx * logits_stride + stop_token_ids,
             -float("inf"),
             mask=mask,
         )
+        if PRESERVE_WHEN_ALL_MASKED:
+            preserve_when_all_masked = tl.load(
+                preserve_when_all_masked_ptr + req_state_idx
+            )
+            if preserve_when_all_masked:
```
Comment (arpera) [inline vllm/v1/worker/gpu/sample/logit_bias.py]: We have `PRESERVE_WHEN_ALL_MASKED` and `preserve_when_all_masked` at the same time. Not the best naming in my humble opinion. Consider renaming it, please.

File: vllm/v1/core/sched/scheduler.py (hunk 6 above)
Comment (Sunt-ing) [reply vllm/v1/core/sched/scheduler.py]: Thanks. I removed the backend-level `grammar.is_terminated()` stop trigger. The `min_tokens` exception now uses vLLM's normal stop-token logic in `check_stop()`, so it still respects `ignore_eos` and user stop tokens. When a step ends on a stop token, the scheduler now drops only that final stop token before advancing the grammar; any preceding accepted tokens are still advanced for validation.

File: vllm/v1/worker/gpu/sample/logit_bias.py (hunk 7 above)
Comment (Sunt-ing) [reply vllm/v1/worker/gpu/sample/logit_bias.py]: Renamed.

**Comment by yzong-rh** [comment]
Hey @Sunt-ing, thanks for the PR! I adapted your repro for Model Runner V1 and V2: https://gist.github.com/yzong-rh/40200cc3831c38f9e79b44182328c08b
Was able to reproduce the issue on main and verified your PR indeed fix this issue.

IIUC correctly there are 3 main changes:
1. Logits processor / sampling side -- ensure some valid token remain unmasked after min_tokens
2. Scheduler side -- add an escape hatch in `check_stop` for min_tokens + structured outputs
3. Structured Outputs side -- prevent FSM advancements on later EOS / stop tokens

I think after the logits processor / sampling is fixed, we won't need fix the other two.
- `check_stop` can return `True` when EOS / stop tokens are sampled.
- FSM advancements shouldn't occur after the first EOS / stop token is encountered.

Here's what it looks like with only the logits processor / sampling fix: https://github.com/vllm-project/vllm/pull/49521. I ran the reproducer on it and it still seems fix the issue. Would you please take a look? Feel free to copy over what makes sense into your PR.

An issue I found is that there's a parity issue between CPU and GPU sampling (MRV1 vs MRV2). On CPU, bad_words is applied before min_tokens. On GPU, logits bias is applied with min_tokens, bad_words is applied after. This is an existing issue however so I don't think it should block this PR, just something to fix later.

**Comment by yzong-rh** [comment]
Also, I think this reveals in a larger issue. Any "semantic masking" logit processor like structured outputs, allowed_token_ids, min_tokens, bad_words, when used in conjunction, can mask out all token_ids (all logits are -inf).

I think ideally we would run these "semantic masking" processors first, check and ensure some token_ids is still allowed, then run the "distribution shaper" processors (e.g. logits_bias, min_p, top_k).

Would require a larger refactor or MRV1 and MRV2 though.

**Comment by vadiklyutiy** [comment]
Just want to highlight that https://github.com/vllm-project/vllm/issues/53665 seems related

**Comment by vadiklyutiy** [comment]
implemented in #54218

**Comment by Sunt-ing** [comment]
Glad to see the issue finally solved.
